### PR TITLE
Separate sale date from publication date

### PR DESCRIPTION
**Pull request by IKarbowiak**
- Add `available_for_purchase` to `Product` model
- Update `Product` type with `available_for_purchase` and `is_available_for_purchase` fields
- Add validations in `checkoutLinesAdd`, `checkoutCreate` and `draftOrderComplete` mutations
- Add validations in `Order.canFinalize` resolver
- Add `setAvailabilityForPurchase` mutation
- Update csv export with `available_for_purchase` field
- Update `populatedb` - 10% of generated products available in future

⚠️ 
We shouldn't merge it before the frontend part is not ready.

# Impact

* [x] New migrations
* [x] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [x] Database migration files are up to date
* [x] The changes are tested
* [x] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/order/utils.py [hunk 1]
```diff
@@ -74,6 +76,22 @@ def validate_product_is_published(order):
             )
 
 
+def validate_product_is_available_for_purchase(order):
+    for line in order:
+        if (
+            not line.variant.product.available_for_purchase
+            or datetime.date.today() < line.variant.product.available_for_purchase
```
Comment (fowczarek) [inline saleor/graphql/order/utils.py:83]: Maby it could be a model function?

File: saleor/graphql/product/mutations/products.py [hunk 2]
```diff
@@ -1817,3 +1818,59 @@ def perform_mutation(cls, _root, info, image_id, variant_id):
...
+                {
+                    "start_date": ValidationError(
+                        "Cannot set start date when is_available is false.",
+                        code=ProductErrorCode.INVALID,
+                    )
+                }
+            )
+
+        if not is_available:
+            product.available_for_purchase = None
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:1869]: Lack of test for this line. :police_officer:

**Review by fowczarek — APPROVED** [review]

**Comment by fowczarek** [comment]
We shouldn't merge it before frontend part is not ready.

**Review by korycins — APPROVED** [review]

**Comment by IKarbowiak** [comment]
Changes in #6060

**Pull request closed without merging**